feat(qna): csv import/export

epaminond opened the pull request.

epaminond committed:
feat(qna): csv import/export

emirotin commented on packages/functionals/botpress-qna/src/index.js:
```diff
@@ -70,5 +74,67 @@ module.exports = {
+    router.get('/csv', async (req, res) => {
+      const questions = await storage.getQuestions()
+      const rows = questions.flatMap(item => {
+        console.log(item)
```
leftover

epaminond replied:
Removed it.

emirotin commented on packages/functionals/botpress-qna/src/index.js:
```diff
@@ -70,5 +74,67 @@ module.exports = {
+
+    router.get('/csv', async (req, res) => {
+      const questions = await storage.getQuestions()
+      const rows = questions.flatMap(item => {
```
do you know since which version Node.js supports `flatMap`?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap says it's not supported but obviously it worked for you. or is it polyfilled in babel-core?

slvnperron replied:
Nice tool for node https://node.green

emirotin commented on packages/functionals/botpress-qna/src/index.js:
```diff
@@ -70,5 +74,67 @@ module.exports = {
+      res.setHeader('Content-disposition', `attachment; filename=qna_${moment().format('DD-MM-YYYY')}.csv`)
+      rows.forEach(({ question, action, answer }) => {
+        res.write(
+          [question, action, answer].map(field => '"' + field.toString().replace(/\"/g, '""') + '"').toString() + '\r\n'
```
This looks fine but maybe it's better to use some external CSV formatter that will handle the escaping for us?

epaminond replied:
Done. Used `json2csv`.

emirotin commented on packages/functionals/botpress-qna/src/index.js:
```diff
@@ -70,5 +74,67 @@ module.exports = {
+
+      if (yn(req.body.isReplace)) {
+        const questions = await storage.getQuestions()
+        await Promise.all(questions.map(({ id }) => storage.deleteQuestion(id)))
```
Can use `Promise.each`

epaminond replied:
I didn't want to add bluebird here, but OK, done.

emirotin commented on packages/functionals/botpress-qna/src/views/index.js:
```diff
@@ -344,18 +348,90 @@ export default class QnaAdmin extends Component {
+                  )}
+                  <form>
+                    <FormGroup>
+                      <ControlLabel>CSV-file</ControlLabel>
```
I'd remove the dash here

emirotin commented on packages/functionals/botpress-qna/src/views/index.js:
```diff
@@ -344,18 +348,90 @@ export default class QnaAdmin extends Component {
+                </Modal.Footer>
+              </Modal>
+
+              <Button bsStyle="success" href="/api/botpress-qna/csv" type="button">
```
It's not protected by any way, right? 
Does this mean it's vulnerable to being downloaded by complete strangers just having the bot URL?

slvnperron replied:
@emirotin @epaminond Could you please check this? By default the `getRouter()` returns a secured router. Not sure with multer what the behavior is but we need to make sure this is secure

emirotin replied:
I mean it's a plan link fetched over GET, how is it secured? It's not getting any headers

slvnperron replied:
Oh you're right. That most certainly doesn't work on a secured dashboard (NODE_ENV=prod). Need to use Axios.

emirotin replied:
Hmmm, does this mean I've recently broken the logs archive download in a recent PR? It went mithe the breaking change so not that bad. Will check and fix today

epaminond replied:
@emirotin , @slvnperron , I've tested it and this route _is_ secured: download only works if user is logged in as admin (the same way it works for other routes in this module).

emirotin replied:
I wonder how it's possible. Unless we're using cookies there's no way to send secret with GET request

epaminond replied:
@emirotin , actually you were right, this button always returned 401 status even when authorized. So I implemented a workaround to download it through axios! Nice catch!

emirotin replied:
Now I need to go back and check the logs stuff I could presumably expose or break :)

slvnperron commented on packages/functionals/botpress-qna/src/index.js:
```diff
@@ -70,5 +74,67 @@ module.exports = {
+
+      try {
+        const content = parse(req.file.buffer.toString(), { columns: ['question', 'action', 'answer'] })
+        await Promise.all(
```
I'd rather not put the logic directly in the router, could you please extract the code to a module that does only one thing?

epaminond replied:
OK, makes sense, moved parsing to separate module.

emirotin commented on packages/functionals/botpress-qna/src/views/index.js:
```diff
@@ -51,6 +54,11 @@ const ACTIONS = {
 export default class QnaAdmin extends Component {
+  constructor(props) {
+    super(props)
+    this.csvDownloadableLink = React.createRef()
```
👍

slvnperron commented on packages/functionals/botpress-qna/src/views/index.js:
```diff
@@ -344,18 +353,107 @@ export default class QnaAdmin extends Component {
+          csvDownloadableLinkHref: window.URL.createObjectURL(new Blob([response.data])),
+          csvDownloadableFileName: /filename=(.*\.csv)/.exec(response.headers['content-disposition'])[1]
+        },
+        () => this.csvDownloadableLink.current.click()
```
CLever

epaminond commented:
@slvnperron , I've implemented exposing APIs here either. I've changed names of some params though to keep them more consistent (e.g. `answer_type` is `action` as it's called in other places). Let me know if you think I should rename them

slvnperron commented:
@epaminond That looks great!!
